### DISCUSSION TITLE
Why use floor divide in shape_div?

### DISCUSSION THREAD

**Discussion by XieXiating** [NVIDIA/cutlass#1770] [Q&A] [answered]
Q: I am using the composition() function and encountered an unexpected result. My input was:

lhs = (_256,(_32,_4)):(_32,(_1,_8192))
rhs = (200,13):(_1@0,_8@1)

However, the result was:

(200,(4,3)):(_32,(_8,_8192))

This is not what I expected. I was expecting:

(200,(4,4)):(_32,(_8,_8192))

I believe the issue arises because the domain<1> of rhs is 13 instead of 12. 

And, I found shape_div() which is used in composition_impl() use floor divided rather than ceil divided.

Can you help clarify this behavior? Thanks

**Comment by ccecka** [comment]
Good catch. At the moment, this is considered a violation of the "divisibility condition" mentioned in the documentation. The static version fails
```
// Fails with shape_div static assertion, indicating a violation of the divisibility condition
composition(make_layout(make_shape(4_s, 4_s), LayoutRight{}),   // (_4,_4):(_4,_1)
            make_layout(13_s));                                 // _13:_1
```

That said, I agree that the divisibility condition is actually too tight in cases like these and the above SHOULD work -- this is a known class of bugs. Fortunately, we have not found any applications that need this generalization, but, unfortunately, supporting it is a bit more complex than simply rounding up. In the near future, I plan to release a much more formal treatment of CuTe in a whitepaper along with some non-critical code updates and generalizations like this one.

**Reply by XieXiating** [reply]
Thanks!

**Reply by ccecka** [reply, marked as answer]
After a cup of coffee on an actual workday, I realize that this particular `composition` case should fail and the current behavior is correct. One post-condition of `composition` is that the result is `compatible` with the rhs input, so `composition` can never perform any rounding at all. Because there is no possible output that satisfies all of the post-conditions of `composition`, it should fail on these inputs (perhaps with better runtime assertions, of course).

There is a related set of known artificial limitations around `composition` and `logical_divide` that can be loosened, but this problem is not an example of them.

**Reply by XieXiating** [reply]
Got it. Thanks for the clarification!
